### PR TITLE
Remove panic from block storage and refactor blockchain checking

### PR DESCRIPTION
**Pull request by tudor-malene**

### CODE REVIEW

File: integration/simulation/stats.go [hunk 1]
```diff
@@ -17,28 +17,28 @@ type Stats struct {
 
 	totalL1Blocks uint64
 
-	totalL2Blocks      uint64
-	l2Head             *nodecommon.Rollup
 	maxRollupsPerBlock uint32
 	nrEmptyBlocks      int
 
-	totalL2Txs  int
+	// totalL2Txs  int
```
Comment (joeldudley) [inline integration/simulation/stats.go:23]: Delete?

**Comment by joeldudley** [comment]
@tudor-malene Do you need to update the go.mod or something? It was complaining about this line in `validate_chain.go` 
until I did a `go get ./...`:

`collect "github.com/sxyazi/go-collection"`

File: integration/simulation/validate_chain.go [hunk 2]
```diff
@@ -22,107 +22,100 @@ type EfficiencyThresholds struct {
...
-	currentBlockHead := obscuroNode.DB().GetCurrentBlockHead()
-	currentRollupHead := obscuroNode.DB().GetCurrentRollupHead()
-
-	l1Height := currentBlockHead.Height
-	l1HeightHash := currentBlockHead.ID
-	l2Height := currentRollupHead.Height
-
-	// ensure the L1 blocks are valid
-	validateL1(t, s.Stats, l1Height, &l1HeightHash, l1Node, efficiencies)
+// After a simulation has run, check as much as possible that everything has
```
Comment (joeldudley) [inline integration/simulation/validate_chain.go:25]: "that everything has" what?

File: integration/simulation/validate_chain.go [hunk 3]
```diff
@@ -22,107 +22,100 @@ type EfficiencyThresholds struct {
...
+func checkNetworkValidity(t *testing.T, s *Simulation, params *SimParams, efficiencies EfficiencyThresholds) {
+	l1MaxHeight := checkEthereumBlockchainValidity(t, s, params, efficiencies)
+	checkObscuroBlockchainValidity(t, s, params, efficiencies, l1MaxHeight)
+}
 
-	// ensure the validity of l1 vs l2 stats
-	validateL1L2Stats(t, obscuroNode, s.Stats, efficiencies)
+// checkEthereumBlockchainValidity: sanity check on the mock implementation of the L1 on all nodes
+// - minimum height - the chain has a minimum number of blocks
+// - check height is similar
```
Comment (joeldudley) [inline integration/simulation/validate_chain.go:33]: maybe `- check height is similar across nodes`?

File: integration/simulation/validate_chain.go [hunk 4]
```diff
@@ -136,100 +129,90 @@ func validateL1(t *testing.T, stats *Stats, l1Height uint64, l1HeightHash *obscu
...
+	totalAmountInSystem := s.Stats.totalDepositedAmount - totalSuccessfullyWithdrawn
+	total := uint64(0)
+	for _, wallet := range s.TxInjector.wallets {
+		total += node.Enclave.Balance(wallet.Address)
+	}
+	if total != totalAmountInSystem {
+		t.Errorf("the amount of money in accounts on node %d does not match the amount deposited. Found %d , expected %d", node.ID, total, totalAmountInSystem)
 	}
+	// TODO Check that processing transactions in the order specified in the list results in the same balances
+	// walk the blocks in reverse direction, execute deposits and transactions and compare to the state in the rollup
```
Comment (joeldudley) [inline integration/simulation/validate_chain.go:204]: Should this also be a todo?

**Review by joeldudley — CHANGES_REQUESTED** [review]
LGTM. A few small things.

**Comment by tudor-malene** [comment]
back to you @joeldudley

**Review by joeldudley — APPROVED** [review]
LGTM

**Pull request merged**